Podcast Player: Add settings colors

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

This PR allows set text and/or background color to the Podcast Player block.


![image](https://user-images.githubusercontent.com/77539/77070087-5ba2e800-69c8-11ea-825b-456589ff92f8.png)

![image](https://user-images.githubusercontent.com/77539/77070072-55ad0700-69c8-11ea-818f-b6548dbac748.png)



Fixes https://github.com/Automattic/jetpack/issues/14940
Depends on https://github.com/Automattic/jetpack/pull/14989

#### Changes proposed in this Pull Request:

* Add `Colors Settings` control to the sidebar.
* Allows set text and background color, either using custom colors or pick them up from the palette defined by the theme.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1) Create/edit a Podcast Player block.
`https://distributed.blog/category/podcast/feed`, just in case ;-)

2) Open and focus block settings.

3) Set text and background color. Every change should be propagated to the block.


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
*

retrofox committed:
[not verified] podcast-player: imp useColors. Check is available

retrofox committed:
[not verified] apply-fallback-styles: create shared HoC

retrofox committed:
[not verified] podcast-player: compose depending on useColors

retrofox committed:
[not verified] podcast-player: implement colors setting pane

retrofox committed:
[not verified] [not verified] podcast-player: add colors function in ss

retrofox committed:
[not verified] [not verified] podcast-player: white list custom color attrs

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: pass attrs to server-side

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: add class_name to get_colors()

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: add color classes to container

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: inherit anchors colors

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: apply color to podcast text

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-player: apply custom colors in server-side

jeryj commented on extensions/blocks/podcast-player/attributes.js:
```diff
@@ -20,4 +20,16 @@ export default {
+	},
+	customBackgroundColor: {
+		type: 'string',
+	},
```
Is using a `customTextColor` normal instead of just using `textColor`? I would have thought that the custom color value would have been applied to `textColor` instead of having a separate value. If this is normal, just dismiss this comment then :)

retrofox replied:
Yes, it's normal. These properties are defined by the `withColors` HoC. You'd see many other implementations in a similar way, even here in extensions.
In short, and just FYI, textColor contains the color defined by a name, for instance, `primary` or something like that. Those 
colors are picked from the color palette, and they are usually defined by the theme.

For instance, `Secondary` is one of the colores fined by twenty-twenty.
<img src="https://user-images.githubusercontent.com/77539/77070577-3cf12100-69c9-11ea-8a07-966ab97e1451.png" width="300px" />

In another hand, custom color is defined by the custom color tool, and its value usually is in RGB:

<img src="https://user-images.githubusercontent.com/77539/77070725-85104380-69c9-11ea-8b65-f4f4da60e092.png" width="300px" />

Depending on how the color was defined, we will apply either CSS classes or inline styles to the HTML markup. SSR also requires some extra handling, too. It isn't a big deal, though.

jeryj commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -49,14 +62,20 @@ const supportUrl =
 const PodcastPlayerEdit = ( {
 	attributes,
 	setAttributes,
+
```
Unnecessary line break

retrofox replied:
I wanted to order attributes by kind. Let me remove these lines :-(

jeryj commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -49,14 +62,20 @@ const supportUrl =
+
 	noticeOperations: { createErrorNotice, removeAllNotices },
 	noticeUI,
+
```
Unnecessary line break

jeryj commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -40,6 +40,18 @@ function register_block() {
+				),
+				'customBackgroundColor'  => array(
+					'type' => 'string',
+				),
```
If the `custom` color comments earlier are relevant, then this would need work too. If this is all normal, then just dismiss my comment :)

retrofox replied:
done!

jeryj commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -179,3 +208,69 @@ function setup_tracks_callback( \SimplePie_Item $episode ) {
+	if ( $has_custom_text_color || $has_named_text_color ) {
+		// Add has-text-color class.
+		$colors['css_classes'][]      = 'has-text-color';
+		$colors['css_text_classes'][] = 'has-text-color';
```
Trying to understand why both of these are necessary. I think `$colors['css_classes'][]      = 'has-text-color';` maybe doesn't need to be there?

retrofox replied:
Very valid comment. We need to know the CSS classes will be applied only for text. It will be used for the podcast title, applied in the `<ol><li>` elements.
For this reason, is why added this extra index to the `$colorsa array.

jeryj commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -179,3 +208,69 @@ function setup_tracks_callback( \SimplePie_Item $episode ) {
+	if ( $has_named_text_color ) {
+		// Add the color class.
+		$colors['css_classes'][]      = sprintf( 'has-%s-color', $attributes['textColor'] );
+		$colors['css_text_classes'][] = sprintf( 'has-%s-color', $attributes['textColor'] );
```
Are both of these necessary? I was trying to read through all of this and it seems a bit complicated to follow the whole flow.

retrofox replied:
yes, we are applying these CSS classes to different HTML elements. The css_classes contains all CSS classes while `css_text_classes` contains only the CSS class about text.

jeryj reviewed (dismissed):
Functionally it seems to work as expected, and nothing stands out to me as an issue.

jeherve commented:
@Copons Would you have some bandwidth to take a look at this today? You've had some experience building color settings for our recent blocks and I think you may have some good insight to share.

Thank you!

retrofox committed:
[not verified] podcast-player: whitelist colors attrs

retrofox committed:
podcast-player: set text color if it isn't defined

retrofox committed:
podcast-player: remove useColor() condition
Let's implement the useColors() implementation for a next PR. For now, it applies settings color as other component, without using the useColors() hook.

retrofox commented:
> @Copons Would you have some bandwidth to take a look at this today? You've had some experience building color settings for our recent blocks and I think you may have some good insight to share.

Some notes about this:

* I've removed the part which tried to deal with using the `useColors()` hack if it was defined. Let's do it in a next PR.
* I've seen duplicated the same code at leas in three blocks. For this reason, I've added the `withFallbackStyles` HoC, and put it into the shared folder. I think we could update the other blocks once it ships 🤞

retrofox committed:
podcast-player: remove line break

Copons commented:
> @Copons Would you have some bandwidth to take a look at this today? You've had some experience building color settings for our recent blocks and I think you may have some good insight to share.

Thanks @jeherve, I'll give this a look, but I think @retrofox has more experience than me, considering lots of my code is _inspired_ by his work on the core Navigation block! 😄

Copons commented:
> * I've removed the part which tried to deal with using the `useColors()` hack if it was defined. Let's do it in a next PR.


> * I've seen duplicated the same code at leas in three blocks. For this reason, I've added the `withFallbackStyles` HoC, and put it into the shared folder. I think we could update the other blocks once it ships 🤞

Nice thinking, very similar to what I did in another PR: https://github.com/Automattic/jetpack/pull/14920/files#diff-3045bc60da5b55bb6a61bcd3f584d7de 🙂 

I think all the other `withFallbackStyles` are used for "button blocks", which are an ongoing pain we are trying to solve once and for all by creating _one last Button component to replace them all_ (#14920).

"Button blocks" are basically blocks that _contain_ a button (which has roughly the same visual attributes of the core Button block), among other elements.
For this reason, in my PR I've opted to change the color attributes, to make it obvious that they refer to the button.
E.g.
`textColor` can mean the button's text color, but also the _rest of the block_ text color.
`buttonTextColor` removes this confusion.

And here's the problem: if we had one single `withFallbackStyles` utility function, we'd also need a way to provide it the names of the color attributes.
Otherwise, we could have _two_ different implementation: yours as a generic one for the default names `textColor` and `backgroundColor`; mine (inside the Button component folder) just for their `button` variations.

Copons commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -218,3 +238,69 @@ function format_track_duration( $duration ) {
+		'style'            => '',
+		'class_name'       => '',
+		'text_class_name'  => '',
+	);
```
I'm a bit unclear about all these return values.

- `class_name` is never used.
- `text_class_name` might be unnecessary. We already return `css_text_classes`, we might as well `implode` it outside of the `get_colors` function.
- What about updating `css_classes` for the background colors only, and then merge it with `css_text_classes` at the end? Something like this:
```php
if ( $has_named_text_color ) {
  $colors['css_text_classes'][] = 'has-text-color';
}
// ...

$colors['css_classes'] = array_merge( $colors['css_classes'], $colors['css_text_classes'] );

return $colors;
```

brbrr commented:
Changed label to "Needs Author Response". Also, I'd like to note some issues found by codeclimate, and WPCOM patch can not apply cleanly. Please, give it a look.

retrofox commented:
> Otherwise, we could have _two_ different implementation: yours as a generic one for the default names `textColor` and `backgroundColor`; mine (inside the Button component folder) just for their `button` variations.

But it isn't for button variations in this case. I think the HoC should as generic as possible.

```es6
/**
 * HoC which will try to pick up the `color` and `backgroundColor`
 * values through DOM manipulation.
 */
```

> E.g.
> textColor can mean the button's text color, but also the rest of the block text color.

That's what we need for this implementation :-D.


I think it doesn't matter if it's used in Buttons, Textareas, etc. Make sense what I say?

marekhrabe commented:
A bit more context about how this will play together with the new `PodcastPlayer` component just merged. We need to compute and prepare everything about colors in the `edit.js` file and then pass colors and classes as props to the `PodcastPlayer`.

The reason is that the `PodcastPlayer` is being used universally in both the editor and the frontend and on the frontend, there will be no context from the editor and there will be no color hooks. It needs to have the final data.

For the frontend to work too, I can see you are already precomputing values like `$block_style` and `$title_classname` - that's great. You can add them to [`$player_data`](https://github.com/Automattic/jetpack/blob/d1f825c3941dba2f49122a12f886647facf010c5/extensions/blocks/podcast-player/podcast-player.php#L98-L102) which will automatically be loaded as props on `PodcastPlayer`.

retrofox commented:
Closing in favor of https://github.com/Automattic/jetpack/pull/15110

retrofox commented:
> A bit more context about how this will play together with the new `PodcastPlayer` component just merged. We need to compute and prepare everything about colors in the `edit.js` file and then pass colors and classes as props to the `PodcastPlayer`.

I prefer to [pass the attributes](https://github.com/Automattic/jetpack/pull/15110/files#diff-ac95db14f00bcb6ae1960269ae9121cfR257) to the `<PodcastPlayer />` instead of changing the data structure. IMO, it makes the implementation transparent, clearer and simpler to understand and develop.

> The reason is that the `PodcastPlayer` is being used universally in both the editor and the frontend and on the frontend, there will be no context from the editor and there will be no color hooks. It needs to have the final data.

It's properly handled propagating the attributes from the parent component to the <PodcastPlayer /> one, [either from `edit`](https://github.com/Automattic/jetpack/pull/15110/files#diff-ac95db14f00bcb6ae1960269ae9121cfR257) or `view`. In `view` case, the attributes are [stringified](https://github.com/Automattic/jetpack/pull/15110/files#diff-551fe0712dea509839f9d2b144489d57R102) and parsed from the server-side.

>
> For the frontend to work too, I can see you are already precomputing values like `$block_style` and `$title_classname` - that's great. You can add them to [`$player_data`](https://github.com/Automattic/jetpack/blob/d1f825c3941dba2f49122a12f886647facf010c5/extensions/blocks/podcast-player/podcast-player.php#L98-L102) ...

It was totally discarded in the current approach. It's better to handle the classes in the front-end in a similar way that the editor-canvas does, especially if we think about the styles applied to the active podcast since it. It isn't possible t do it n the server-side.
The previous code made more sense when the SSR was still there, before to the react for universal rendering.

...which will automatically be loaded as props on `PodcastPlayer`.

In the current PR, you can pick up the attributes instead of thinking on props, which makes the universality of the component works better.
